vignettes/data-as-shiny-module.Rmd

```diff
@@ -0,0 +1,127 @@
+---
+title: "Data as shiny Module"
+author: "NEST CoreDev"
+output: 
+  rmarkdown::html_vignette:
+    toc: true
+vignette: >
+  %\VignetteIndexEntry{Data as shiny Module}
+  %\VignetteEngine{knitr::rmarkdown}
+  %\VignetteEncoding{UTF-8}
+---
+
+## Introduction
+
+Proper functioning of any `teal` application requires presence of a `teal_data` object.
+Typically, a `teal_data` object created in the global environment will be passed to the `data` argument in `init`.
```

Comment:
Does it have to be a _global env_? I propose to rephrase it into just: It's required as a `data` argument in `init`.

Reply:
There are few more occurrences of "global environment". I feel that what we want to say is: "before the app defition" (or rephrase of this). We can include following article to better understand the difference: https://shiny.posit.co/r/articles/improve/scoping/

UPDATE: actually we can re-use the terminology from there which is: "global objects" / "per-session objects".

Reply:
Global environment is still a correct, precise term. "global objects" is an informal name used by `shiny` but they still explain as

```r
Objects defined in global.R (...) they are loaded into the global environment of the R session
```

We can say also, "objects created before `teal::init`"?

Reply:
I actually find other, much simpler statement that I find useful here:

> outside of the server function definition

I think I can do the following
```
e <- new_env()
e$x <- teal_data(...)
init(data = e$x, ...)
```

Whether `e$x` is a global env object or not is debatable but IMHO it fulfils the definition of a "global object".

Or the following:
```
init(data = teal_data(...), ...)
```

No assignment to global env.

Maybe we can have a short glossary at the very beginning? My point is ~not to use~ limit the use an "environment" term throughout the content of this article as this might be too complex for an average user. My idea is to refer that article first and say that teal-data-module is an implementation of per-session data objects whereas global objects are still allowed (see other vignette here: XYZ).

Reply:
I agree that we don't want to burden the app dev with environments but "global environment" is a basic concept of interactive R.

I don't think a glossary would solve the problem. You would have to explain it in terms of environments anyway. Plus it's clutter, better to use footnotes.

> I think I can do the following
> 
> ```
> e <- new_env()
> e$x <- teal_data(...)
> init(data = e$x, ...)
> ```

You _can_ but would you? It seems a little far-fetched. Please note that the vignette says "_Typically_, a `teal_data` object created in the global environment will be passed to the `data` argument in `init`."

Reply:
yeah right - don't want to replace technicality with a clutter

My point is to make it more _simpler_ and avoid frequent use of "environment". Thought we can "delegate" that problem to the aforementioned article. If you think it's ok then let's leave it.

Reply:
If we don't have a clear conclusion I suggest to leave it aside and open an issue to revisit this vignette. This PR blocks a lot and each day postpones even bigger issues in teal which we have to solve in 1 month.

Reply:
> If we don't have a clear conclusion I suggest to leave it aside and open an issue to revisit this vignette. This PR blocks a lot and each day postpones even bigger issues in teal which we have to solve in 1 month.

I agree. We can seek an outsider's opinion in the mean time.

Reply:
I'm ok with this. My vignettes comments are not that strong anyway.